### PR TITLE
Upgrade React Native to 0.41

### PR DESCRIPTION
**Pull request by hawkrives**
> closes #483 
> closes #656 
> closes #674 
> closes #681 
> closes #685 

Upgrade React Native to 0.41!

This also upgrades all our dependencies.

The giant diff here is because we no longer need to use CocoaPods to include OneSignal; instead, `react-native-onesignal` has opted to ship the CocoaPods source within their npm package, just like every other library we use.

This allowed us to get rid of CocoaPods entirely, so from now on you will need to open the `.xcodeproj` file instead of the `.xcworkspace` file.

Feel free to ask questions about any of the changes in here.

It passes CI, so I'll merge it before tonight's beta and we'll see if anything breaks in there I guess.

You might need to / want to run this command to update your build-tools:

Windows:
```bat
$ echo y | "%ANDROID_HOME%\tools\android.bat" --silent update sdk --no-ui --all --filter android-23,platform-tools,tools,build-tools-23.0.1,build-tools-23.0.2,extra-google-m2repository,extra-android-m2repository,extra-android-support
```

Linux/Mac:
```console
$ echo y | android --silent update sdk --no-ui --all --filter android-23,platform-tools,tools,build-tools-23.0.1,build-tools-23.0.2,extra-google-m2repository,extra-android-m2repository,extra-android-support
```

### CODE REVIEW

**Comment by hawkrives** [comment]
OneSignal completely changed how it's integrated with react-native, so I removed it. I will be adding it back again.

**Comment by hawkrives** [comment]
oh hai danger

**Comment by hawkrives** [comment]
Alright, it's green. I'm going to update the OP to actually have details of what changed.

File: .flowconfig [hunk 1]
```diff
@@ -45,15 +45,14 @@ module.file_ext=.js
 module.file_ext=.jsx
 module.file_ext=.json
 
-module.name_mapper='^image![a-zA-Z0-9$_-]+$' -> 'GlobalImageStub'
```
Comment (hawkrives) [inline .flowconfig:48]: React deprecated `require('image!./img.png')` support a long time ago, and they just completely removed it.

File: .gitignore [hunk 2]
```diff
@@ -42,7 +43,6 @@ android/app/libs
 *.keystore
 
 # KEYS
-keys.js
```
Comment (hawkrives) [inline .gitignore:45]: Removed ignore for `keys.js`, since we haven't used it as a file since September.

File: .travis.yml [hunk 3]
```diff
@@ -38,6 +38,7 @@ matrix:
           - tools
           - platform-tools
           - build-tools-23.0.1
+          - build-tools-23.0.2
```
Comment (hawkrives) [inline .travis.yml:41]: Added `build-tools-23.0.2` for faster builds

File: android/app/build.gradle [hunk 4]
```diff
@@ -140,19 +140,21 @@ android {
...
-    compile project(':react-native-vector-icons')
+    compile project(':react-native-google-analytics-bridge')
     compile project(':react-native-keychain')
+    compile project(':react-native-onesignal')
+    compile project(':react-native-vector-icons')
+    // this is for react-native itself
     compile fileTree(dir: "libs", include: ["*.jar"])
     compile "com.android.support:appcompat-v7:23.0.1"
     compile "com.facebook.react:react-native:+"  // From node_modules
-    compile project(':react-native-onesignal')
```
Comment (hawkrives) [inline android/app/build.gradle:150]: I sorted these alphabetically

File: android/app/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -13,11 +13,11 @@
         android:targetSdkVersion="22" />
 
     <application
+      android:name=".MainApplication"
       android:allowBackup="true"
       android:label="@string/app_name"
       android:icon="@mipmap/ic_launcher"
-      android:theme="@style/AppTheme"
-      android:name=".MainApplication">
+      android:theme="@style/AppTheme">
```
Comment (hawkrives) [inline android/app/src/main/AndroidManifest.xml:20]: This is how React-Native wants to sort this

File: android/app/src/main/java/com/allaboutolaf/MainApplication.java [hunk 6]
```diff
@@ -3,17 +3,18 @@
...
 import com.facebook.react.ReactNativeHost;
 import com.facebook.react.ReactPackage;
 import com.facebook.react.shell.MainReactPackage;
-import com.geektime.reactnativeonesignal.ReactNativeOneSignalPackage;
 import com.facebook.soloader.SoLoader;
+import com.geektime.rnonesignalandroid.ReactNativeOneSignalPackage;
+import com.idehub.GoogleAnalyticsBridge.GoogleAnalyticsBridgePackage;
+import com.learnium.RNDeviceInfo.RNDeviceInfo;
+import com.oblador.keychain.KeychainPackage;
+import com.oblador.vectoricons.VectorIconsPackage;
```
Comment (hawkrives) [inline android/app/src/main/java/com/allaboutolaf/MainApplication.java:17]: I sorted these alphabetically

File: android/app/src/main/java/com/allaboutolaf/MainApplication.java [hunk 7]
```diff
@@ -22,19 +23,20 @@
 
   private final ReactNativeHost mReactNativeHost = new ReactNativeHost(this) {
     @Override
-    protected boolean getUseDeveloperSupport() {
+    public boolean getUseDeveloperSupport() {
```
Comment (hawkrives) [inline android/app/src/main/java/com/allaboutolaf/MainApplication.java:26]: RN did this

File: android/app/src/main/java/com/allaboutolaf/MainApplication.java [hunk 8]
```diff
@@ -22,19 +23,20 @@
...
         new MainReactPackage(),
+        // please keep these sorted alphabetically
         new GoogleAnalyticsBridgePackage(),
-        new RNDeviceInfo(),
-        new VectorIconsPackage(),
         new KeychainPackage(),
-        new ReactNativeOneSignalPackage()
+        new ReactNativeOneSignalPackage(),
+        new RNDeviceInfo(),
+        new VectorIconsPackage()
```
Comment (hawkrives) [inline android/app/src/main/java/com/allaboutolaf/MainApplication.java:39]: I sorted these alphabetically

File: android/build.gradle [hunk 9]
```diff
@@ -5,7 +5,7 @@ buildscript {
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.1.3'
+        classpath 'com.android.tools.build:gradle:2.2.2'
```
Comment (hawkrives) [inline android/build.gradle:8]: Updated this for faster builds

File: android/settings.gradle [hunk 10]
```diff
@@ -1,13 +1,15 @@
...
+include ':react-native-google-analytics-bridge'
+project(':react-native-google-analytics-bridge').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-google-analytics-bridge/android')
 include ':react-native-keychain'
 project(':react-native-keychain').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-keychain/android')
 include ':react-native-onesignal'
 project(':react-native-onesignal').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-onesignal/android')
-include ':react-native-google-analytics-bridge'
-project(':react-native-google-analytics-bridge').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-google-analytics-bridge/android')
+include ':react-native-vector-icons'
+project(':react-native-vector-icons').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-vector-icons/android')
```
Comment (hawkrives) [inline android/settings.gradle:15]: I sorted these alphabetically

File: app.js [hunk 11]
```diff
@@ -7,6 +7,7 @@
 // Tweak the global fetch
 import './globalize-fetch'
 import {tracker} from './analytics'
+import OneSignal from 'react-native-onesignal'
```
Comment (hawkrives) [inline app.js:10]: OneSignal 2/3 moved the OneSignal config into app.js

File: app.json [hunk 12]
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "AllAboutOlaf",
+  "displayName": "All About Olaf"
+}
```
Comment (hawkrives) [inline app.json:4]: I'm still not entirely clear on the purpose of this file – I think it's optional, but the upgrade added it. We'll see what it's used for in the future.

File: fastlane/Fastfile [hunk 13]
```diff
@@ -42,6 +42,11 @@ platform :ios do
     snapshot
   end
 
+  desc "Go rogue"
+  lane :rogue do
+    go_rogue
+  end
+
```
Comment (hawkrives) [inline fastlane/Fastfile:49]: Added `fastlane ios rogue` for the rogue team

File: fastlane/Fastfile [hunk 14]
```diff
@@ -108,19 +111,6 @@ platform :ios do
...
-
-    # append the HockeySDK to the Podfile, unless it already lists it
-    File.open(podpath, "a") do |file|
-      file << "pod \"HockeySDK\"\n"
-    end unless File.foreach(podpath).grep(/HockeySDK/).any?
-
-    # and install the new pods
-    cocoapods(podfile: "./ios")
-  end
-
```
Comment (hawkrives) [inline fastlane/Fastfile:123]: Opted to remove the HockeySDK since we don't have Cocoapods anymore

File: fastlane/Gymfile [hunk 15]
```diff
@@ -1,7 +1,7 @@
 # For more information about this configuration visit
 # https://github.com/fastlane/fastlane/tree/master/gym#gymfile
 
-workspace "ios/AllAboutOlaf.xcworkspace"
+project "ios/AllAboutOlaf.xcodeproj"
```
Comment (hawkrives) [inline fastlane/Gymfile:4]: No more cocoapods! No more workspace!

File: ios/Podfile [hunk 16]
```diff
@@ -1,9 +0,0 @@
-target 'AllAboutOlaf' do
-  # Pods for AllAboutOlaf
-  pod 'OneSignal', '~> 1.13.2'
-
-  target 'AllAboutOlafTests' do
-    inherit! :search_paths
-    # Pods for testing
-  end
-end
```
Comment (hawkrives) [inline ios/Podfile:9]: Removed Cocoapods because react-native-onesignal changed how it loads OneSignal

File: views/settings/index.js [hunk 17]
```diff
@@ -49,8 +49,8 @@ function SettingsView(props: SettingsViewPropsType) {
   return (
     <ScrollView
       contentContainerStyle={styles.container}
-      keyboardShouldPersistTaps={true}
-      keyboardDismissMode={'on-drag'}
+      keyboardShouldPersistTaps='always'
```
Comment (hawkrives) [inline views/settings/index.js:52]: rn@0.41 changed this from a boolean to a string

File: lib/login.js [hunk 18]
```diff
@@ -1,5 +1,5 @@
 // @flow
-import Keychain from 'react-native-keychain'
+import * as Keychain from 'react-native-keychain'
```
Comment (hawkrives) [inline lib/login.js:2]: rnKeychain@1.0 changed how they export

File: lib/formdata.js [hunk 19]
```diff
@@ -2,8 +2,10 @@
 
 export default function buildFormData(obj: {[key: string]: string}): FormData {
   let formData = new FormData()
-  for (let [key, val] of Object.entries(obj)) {
-    formData.append(key, val)
+  for (const key in obj) {
+    if (obj.hasOwnProperty(key)) {
+      formData.append(key, obj[key])
+    }
```
Comment (hawkrives) [inline lib/formdata.js:8]: Flow is silly and has a weird for Object.entries()

File: app.json (hunk 12 above)
Comment (drewvolz) [reply app.json:4]: I had to do some digging. I think it's part of a CLI to more easily recreate native folders? https://github.com/facebook/react-native/pull/12162

File: fastlane/Fastfile (hunk 13 above)
Comment (drewvolz) [reply fastlane/Fastfile:49]: 👍

File: fastlane/Gymfile (hunk 15 above)
Comment (drewvolz) [reply fastlane/Gymfile:4]: Woo!

File: app.js (hunk 11 above)
Comment (drewvolz) [reply app.js:10]: That is a welcomed improvement.

File: lib/login.js (hunk 18 above)
Comment (drewvolz) [inline lib/login.js]: Something is up with the keychain. I'm seeing an error in `St. Olaf Login`.

File: app.json (hunk 12 above)
Comment (hawkrives) [reply app.json:4]: I thought so too, but that was only introduced eight days ago? It's not in _any_ tags yet.

**Comment by drewvolz** [comment]
Something is up with the keychain. I'm seeing an error when I load the app on my physical device. 

> "Argument 1 (RCTResponseSenderBlock) of RNKeychainManager. must not be null"

**Comment by drewvolz** [comment]
@hawkrives concluded that I was running a rn-39 binary against the rn-41 branch. An error I had is not an error any longer.

**Review by drewvolz — APPROVED** [review]

**Pull request merged**